### DISCUSSION TITLE
Confused about Scoped registrations in Prism IoC

### DISCUSSION THREAD

**Discussion by bzuillsmith** [PrismLibrary/Prism#2655] [General] [answered]
I'm trying to understand scopes in prism Ioc. I have a program that needs to connect to an arbitrary db file and then create a bunch of data services and run some operations. After it's done, I throw them away. I think I would do this with scoped registrations? But I'm struggling to understand how to do that exactly without casting everything to concrete container classes and working with them directly.

I looked at the tests to understand scopes via prism IoC and now I'm more confused. In this example below, `CreateScope()` is called but the result is not used. It looks like it's calling `Resolve<>()` on the same parent container? The code looks like CreateScope() creates a child container. It seems like I would register my DB connection object onto this child but the IScopedProvider doesn't have registration functions in its interface.

https://github.com/PrismLibrary/Prism/blob/342689ba9a10e511418943438d0f5c57534ca623/tests/Containers/Prism.Container.Shared/Tests/ContainerFixture.cs#L355-L365

Can someone clarify how scope/child containers work in Prism Ioc and how I might best go about registering my temporary connection objects?

**Comment by bzuillsmith** [comment]
After struggling with this a bit more, I'm starting to understand that scoped registrations are different from registering things on a child container. It does not appear Prism has an interface for doing registrations on a child at runtime. Instead I have to grab the underlying container with a cast, create a child, and register my connection object. Was hoping to avoid it so I can port it easier in the future to new versions or different container implementations, but it's not a ton of code in my case so it should be fine.

**Reply by ssingh-cemtrex** [reply]
Unfortunately I am still struggling. Can you please elaborate the solution ?

**Comment by dansiegel** [comment, marked as answer]
Scoping in Prism is only implemented within Prism.Forms. In this case the scope is set to the Page. Each Page has its own scope. For WPF & Uno we do not deal with creating scopes and this becomes the burden of the developer. 

Prism does not implement registrations on a child container as this goes against our established patterns and recommendations. You should configure the container at application startup / via Prism Modules. 

```cs
containerRegistry.RegisterScoped<IFoo, Foo>();
```

For context you should consider the following when trying to determine what registration is best.
1. Does my service need to exist and be the SAME EXACT instance everywhere in my application? Then it should be a Singleton
2. I only use this service in a few places, it doesn't need to sit in memory as it can be a new instance when I need it.. Then it should be a Transient.
3. I need to use this service in ServiceA and ServiceB and it should be the same instance but that's really based on the Context of those 2 services being injected into my ViewModel and I don't need it in other ViewModels... then it could be Scoped.

> In this example below, CreateScope() is called but the result is not used. It looks like it's calling Resolve<>() on the same parent container?

If you look closely the Container Extension locally stores the latest scope and resolves from it. Additionally you'll notice that we do store off the scope in the PrismNavigationService so that we can attach the scope to the Page.

**Reply by bzuillsmith** [reply]
Thanks for the clear explanation and filling in some missing details for me. As I suspected, my situation doesn't really fall into 1, 2, or 3. Makes sense though. What I'm doing is a bit beyond what the framework is meant to handle since it doesn't directly relate to UI.
